service/http.proxy/src/test/java/org/kaazing/gateway/service/http/proxy/HttpProxySecurePersistenceTest.java

```diff
@@ -0,0 +1,172 @@
+    private final KeyStore trustStore = TlsTestUtil.trustStore();
+    private final SSLSocketFactory clientSocketFactory = TlsTestUtil.clientSocketFactory();
+
+    private static final int KEEP_ALIVE_TIMEOUT = 5;
+
+    @Rule
+    public TestRule timeout = new DisableOnDebug(new Timeout(15, SECONDS));
+
+    @Test
+    public void securePersistentConnection() throws Exception {
+        Gateway gateway = new Gateway();
+        // @formatter:off
+        GatewayConfiguration configuration =
+                new GatewayConfigurationBuilder()
+                    .service()
+                        .accept(URI.create("https://localhost:8110"))
+                        .connect(URI.create("https://localhost:8080"))
+                        .type("http.proxy")
+                        .connectOption("http.keepaliveTimeout", String.valueOf(KEEP_ALIVE_TIMEOUT))
+                    .done()
+                    .security()
+                        .trustStore(trustStore)
+                        .keyStore(keyStore)
+                        .keyStorePassword(password)
+                    .done()
+                .done();
+        // @formatter:on
+
+        SecureOriginServer.Handler handler = new HttpServer();
+        SecureOriginServer originServer = new SecureOriginServer(8080, handler);
+
+        try {
+            originServer.start();
+            gateway.start(configuration);
+
+            Socket socket = clientSocketFactory.createSocket("localhost", 8110);
+            HttpClient client = new HttpClient();
+            client.handle(socket);
+        } finally {
+            gateway.stop();
+            originServer.stop();
+        }
+
+    }
+
+    private static class HttpClient {
+        static final byte[] HTTP_REQUEST =
+                ("GET / HTTP/1.1\r\n" +
+                "Host: localhost:8110\r\n" +
+                "Cache-Control: max-age=0\r\n" +
+                "Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8\r\n" +
+                "User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.6; rv:8.0) Gecko/20100101 Firefox/8.0\r\n" +
+                "Accept-Encoding: gzip, deflate, sdch\r\n" +
+                "Accept-Language: en-US,en;q=0.8\r\n" +
+                "\r\n").getBytes(UTF_8);
+
+
+        void handle(Socket clientSocket) throws IOException {
+            try (Socket socket = clientSocket;
+                 InputStream in = socket.getInputStream();
+                 OutputStream out = socket.getOutputStream()) {
+
+                for(int i=0; i < 3; i++) {
+                    // read and write HTTP request and response headers
+                    out.write(HTTP_REQUEST);
+                    SecureOriginServer.HttpHandler.parseHttpHeaders(in);
+                    readFully(in, new byte[31]);
+                }
+
+                // sleep so that server persistent connection times out
+                try {
+                    Thread.sleep(KEEP_ALIVE_TIMEOUT*1000 + 3);
```

Comment:
Since InputStream is blocking, it doesn't return -1 unless it is closed. If there is no data, the read() is just blocked. If read() throws an exception, the test should fail (as originServer.stop() would throw that exception)